src/hazelweb/model/Program.re

```diff
@@ -91,6 +91,249 @@ let get_decoration_paths = (program: t): UHDecorationPaths.t => {
+    let hii = HoleInstanceInfo.update_environment(hii, (u, i), sigma);
+    (EmptyHole(u, i, sigma), hii);
+  | NonEmptyHole(reason, u, i, sigma, d1) =>
+    let (sigma, hii) = renumber_sigma(path, u, i, hii, sigma);
+    let hii = HoleInstanceInfo.update_environment(hii, (u, i), sigma);
+    let (d1, hii) = renumber_sigmas_only(path, hii, d1);
+    (NonEmptyHole(reason, u, i, sigma, d1), hii);
+  | FreeVar(u, i, sigma, x) =>
+    let (sigma, hii) = renumber_sigma(path, u, i, hii, sigma);
+    let hii = HoleInstanceInfo.update_environment(hii, (u, i), sigma);
+    (FreeVar(u, i, sigma, x), hii);
+  | Keyword(u, i, sigma, k) =>
+    let (sigma, hii) = renumber_sigma(path, u, i, hii, sigma);
+    let hii = HoleInstanceInfo.update_environment(hii, (u, i), sigma);
+    (Keyword(u, i, sigma, k), hii);
+  | Cast(d1, ty1, ty2) =>
+    let (d1, hii) = renumber_sigmas_only(path, hii, d1);
+    (Cast(d1, ty1, ty2), hii);
+  | FailedCast(d1, ty1, ty2) =>
+    let (d1, hii) = renumber_sigmas_only(path, hii, d1);
+    (FailedCast(d1, ty1, ty2), hii);
+  | InvalidOperation(d, err) =>
+    let (d, hii) = renumber_sigmas_only(path, hii, d);
+    (InvalidOperation(d, err), hii);
+  }
+and renumber_sigmas_only_rules =
+    (path: InstancePath.t, hii: HoleInstanceInfo.t, rules: list(DHExp.rule))
+    : (list(DHExp.rule), HoleInstanceInfo.t) =>
+  rules
+  |> List.fold_left(
+       (b, r: DHExp.rule) => {
+         let (rs, hii) = b;
+         switch (r) {
+         | Rule(dp, d) =>
+           /* pattern holes don't have environments */
+           let (d, hii) = renumber_sigmas_only(path, hii, d);
+           (rs @ [DHExp.Rule(dp, d)], hii);
+         };
+       },
+       ([], hii),
+     )
+and renumber_sigma =
+    (
+      path: InstancePath.t,
+      u: MetaVar.t,
+      i: MetaVarInst.t,
+      hii: HoleInstanceInfo.t,
+      sigma: Environment.t,
+    )
+    : (Environment.t, HoleInstanceInfo.t) => {
+  let (sigma, hii) =
+    List.fold_right(
+      (xd: (Var.t, DHExp.t), acc: (Environment.t, HoleInstanceInfo.t)) => {
+        let (x, d) = xd;
+        let (sigma_in, hii) = acc;
+        let path = [((u, i), x), ...path];
+        let (d, hii) = renumber_result_only(path, hii, d);
+        let sigma_out = [(x, d), ...sigma_in];
+        (sigma_out, hii);
+      },
+      sigma,
+      ([], hii),
+    );
+
+  List.fold_right(
+    (xd: (Var.t, DHExp.t), acc: (Environment.t, HoleInstanceInfo.t)) => {
+      let (x, d) = xd;
+      let (sigma_in, hii) = acc;
+      let path = [((u, i), x), ...path];
+      let (d, hii) = renumber_sigmas_only(path, hii, d);
+      let sigma_out = [(x, d), ...sigma_in];
+      (sigma_out, hii);
+    },
+    sigma,
+    ([], hii),
+  );
+};
+
+let renumber =
+    (path: InstancePath.t, hii: HoleInstanceInfo.t, d: DHExp.t)
+    : (DHExp.t, HoleInstanceInfo.t) => {
+  let (d, hii) = renumber_result_only(path, hii, d);
+  renumber_sigmas_only(path, hii, d);
+};
+
 exception DoesNotElaborate;
 let expand =
```

Comment:
this should be renamed to elaborate (similarly for other uses of "expand" / "expansion")